### PR TITLE
[move-prover] Cleaning prelude and experimenting with vector theories.

### PR DESCRIPTION
**Pull request by wrwg**
This cleans up the prelude (in the boogie v2 backend, guarded by flag `--vnext`), pulling out general data structures like vectors and multisets into independent files and making them generic using Boogie's new monomorphization feature. It also supports now multiple theories for vectors with and without extensionality using the new `--vector-theory` flag. However, only the style of theory we had until now (Boogie arrays wo/ extensional equality) works well in all scenarios.

The supported theories are:

- `--vector-theory=BoogieArray` (default): similar like old `$ValueArray` (now type `Vec $Value`).
- `--vector-theory=BoogieArrayIntern`: attempt to make vectors extensional by introducing an internalization function s.t. `$IsEqual(v1, v2) <==> $Intern(v1) == $Intern(v2)`. Works semantically but times out on simple examples like `simple_vector_client.move`.
- `--vector-theory=SmtSeq`: uses smt sequence types. Too slow and times out on basic examples in `simple_vector_client.move` test.
- `--vector-theory=SmtArray`: use native array theory instead of Boogie array theory. Behaves mostly similar regards timing than `BoogieArray` but runs very long on some modules. The cases where it runs long may be well caused by the butterfly effect (DiemSystem add and remove validator).
- `--vector-theory=SmtArrayExt`: use native array theory and adds axioms for extensionality, as done also with Boogie `-lib` flag. Times out on simple examples.

There is also a new script `tools/benchmark.sh` for recording verification times of a given set of modules, and some of its results in `tools/data`.

This PR cannot be merged before the recent fixes to Boogie monomorphization have been released and the Boogie version is updated.

## Motivation

Investigating whether we can have an encoding with extensional equality (result: no for now)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Existing tests.

## Related PRs

NA

## If targeting a release branch, please fill the below out as well

NA

### CODE REVIEW

**Comment by wrwg** [comment]
@barrettcw  FYI, after this PR landed, we should be able to reactivate the CVC4 tests. It should also be possible to easily swap the theory of vectors since we now have a clean API (see [./boggie-backend-v2/src/prelude](https://github.com/wrwg/diem/tree/extensionality/language/move-prover/boogie-backend-v2/src/prelude). 

This PR showed to me is that none of the alternatives we have been looking at (native smt arrays with extensionality axioms, smt sequences, etc.) is anywhere close to work as well for us as the Boogie arrays with stratified equality. Typically our so-called [simple-vector-test.move](https://github.com/wrwg/diem/blob/extensionality/language/move-prover/tests/sources/functional/simple_vector_client.move) is already failing. 

My intention is to make the integration of other solvers and different theories generally simpler to facilitate better experimentation.

**Comment by shazqadeer** [comment]
Fixes to Boogie monomorphization have been released in 2.8.29 which is now public available.

**Comment by wrwg** [comment]
FYI, one aspect of this PR which came as a side-effect is some very first attempt effort to [preserve benchmark data](https://github.com/wrwg/diem/tree/extensionality/language/move-prover/tools/data).

**Comment by wrwg** [comment]
> Fixes to Boogie monomorphization have been released in 2.8.29 which is now public available.

Thanks! Updated the PR to upgrade to this version, and now testing both default and `--vnext` based on monomorphization on CI.

**Comment by barrettcw** [comment]
This is really great!  It will be much easier to experiment with different backends now.

File: language/move-prover/boogie-backend-v2/src/boogie_wrapper.rs [hunk 1]
```diff
@@ -612,21 +609,15 @@ pub enum ValueArrayRep {
...
     fn new(wrapper: &BoogieWrapper<'_>) -> Self {
         Model {
             vars: Default::default(),
-            tracked_exps: Default::default(),
-            value_array_rep: if wrapper.options.vector_using_sequences {
-                ValueArrayRep::ValueArrayIsSeq
-            } else {
-                ValueArrayRep::ValueArrayIsMap
-            },
+            vector_theory: wrapper.options.vector_theory,
```
Comment (shazqadeer) [inline language/move-prover/boogie-backend-v2/src/boogie_wrapper.rs:620]: General question:  Are you supporting model extraction and visualization for all vector representations?

File: language/move-prover/boogie-backend-v2/src/boogie_wrapper.rs [hunk 2]
```diff
@@ -723,16 +674,10 @@ impl ModelValue {
...
-    /// to extract the actual values.
-    fn extract_vector(&self, model: &Model) -> Option<ModelValueVector> {
-        // In current encoding, directly forward to value array,
-        self.extract_value_array(model)
-    }
-
-    /// Extracts a value array from it's representation.
-    /// If the representation uses maps it is defined by `(ValueArray map_key size)`. The function
-    /// follows indirections in the model. We find the value array map at `Select_[$int]$Value`.
+    /// Extracts a vector from it's representation.
```
Comment (shazqadeer) [inline language/move-prover/boogie-backend-v2/src/boogie_wrapper.rs:677]: it's -> its

File: language/move-prover/boogie-backend-v2/src/prelude/multiset-array-theory.bpl [hunk 3]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+// Boogie model for multisets, based on Boogie arrays. This theory assumes extensional equality for element types.
+
+type {:datatype} Multiset _;
```
Comment (shazqadeer) [inline language/move-prover/boogie-backend-v2/src/prelude/multiset-array-theory.bpl:6]: I suppose this one is being used for Move events.

File: language/move-prover/boogie-backend-v2/src/prelude/vector-array-intern-theory.bpl [hunk 4]
```diff
@@ -0,0 +1,177 @@
+// Copyright (c) The Diem Core Contributors
```
Comment (shazqadeer) [inline language/move-prover/boogie-backend-v2/src/prelude/vector-array-intern-theory.bpl:1]: It is really nice to see different encodings in different files.

File: language/move-prover/boogie-backend-v2/src/prelude/vector-array-theory.bpl [hunk 5]
```diff
@@ -0,0 +1,155 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+// Basic theory for vectors using arrays. This version of vectors is not extensional.
+
+type {:datatype} Vec _;
+
+function {:constructor} Vec<T>(v: [int]T, l: int): Vec T;
+
+function {:builtin "MapConst"} MapConstVec<T>(T): [int]T;
```
Comment (shazqadeer) [inline language/move-prover/boogie-backend-v2/src/prelude/vector-array-theory.bpl:10]: I just want to point out that `:builtin "MapConst"` is interpreted by Boogie only if /useArrayTheory is passed which indicates to Boogie that it should the native SMT array theory.

File: language/move-prover/tests/testsuite.rs [hunk 6]
```diff
@@ -33,9 +33,56 @@ const REGULAR_TEST_FLAGS: &[&str] = &[
 
 static NOT_CONFIGURED_WARNED: AtomicBool = AtomicBool::new(false);
 
-fn test_runner(path: &Path) -> datatest_stable::Result<()> {
+// A list of different feature variants to test. Each test will be executed for each variant,
+// unless it (a) contains a line comment of the form `// exclude_for: <feature1> <feature2> ...` (b)
+// its path is in the blacklisted paths specified.
+const TESTED_FEATURES: &[&str] = &["default", "vnext"];
+const BLACKLISTED_PATHS: &[(&str, &str)] = &[];
```
Comment (shazqadeer) [inline language/move-prover/tests/testsuite.rs:40]: BLACKLISTED -> DENYLISTED

File: language/move-prover/boogie-backend-v2/src/prelude/vector-smt-array-theory.bpl [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+// Boogie model for vectors, based on smt arrays.
+
+// This version of vectors requires boogie to be called with `-useArrayTheory`.
+// It is not extensional.
+
+// Currently we just include the basic vector array theory.
+
+{{> vector-array-theory }}
```
Comment (shazqadeer) [inline language/move-prover/boogie-backend-v2/src/prelude/vector-smt-array-theory.bpl:11]: What does this mean?

Comment (shazqadeer) [reply language/move-prover/boogie-backend-v2/src/prelude/vector-smt-array-theory.bpl:11]: Ah, I see.  This just includes another file (which is being included in more than one place, hence the indirection).

File: language/move-prover/boogie-backend-v2/src/prelude/vector-array-intern-theory.bpl [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+function {:constructor} Vec<T>(v: [int]T, l: int): Vec T;
+
+function {:builtin "MapConst"} MapConstVec<T>(T): [int]T;
+function {:inline} DefaultVecElem<T>(): T;
+function {:inline} DefaultVecMap<T>(): [int]T { MapConstVec(DefaultVecElem()) }
+
+axiom {:ctor "Vec"} (forall<T> x: Vec T :: {l#Vec(x)} l#Vec(x) >= 0);
+
+function {:inline} VecIntern<T>(v: Vec T): Vec T;
```
Comment (shazqadeer) [inline language/move-prover/boogie-backend-v2/src/prelude/vector-array-intern-theory.bpl:19]: :inline meaningful only on a function with a body.

**Review by shazqadeer — APPROVED** [review]

File: language/move-prover/boogie-backend-v2/src/boogie_wrapper.rs (hunk 1 above)
Comment (wrwg) [reply language/move-prover/boogie-backend-v2/src/boogie_wrapper.rs:620]: Only BoogieArray and SmtSeq for now.

File: language/move-prover/boogie-backend-v2/src/prelude/multiset-array-theory.bpl (hunk 3 above)
Comment (wrwg) [reply language/move-prover/boogie-backend-v2/src/prelude/multiset-array-theory.bpl:6]: Yes

File: language/move-prover/boogie-backend-v2/src/prelude/vector-array-theory.bpl (hunk 5 above)
Comment (wrwg) [reply language/move-prover/boogie-backend-v2/src/prelude/vector-array-theory.bpl:10]: Interesting, we have been using it before like this for ValueArray. I guess it doesn't matter for the non-extensional theories.

File: language/move-prover/tests/testsuite.rs (hunk 6 above)
Comment (wrwg) [reply language/move-prover/tests/testsuite.rs:40]: Thanks for pointing out.

File: language/move-prover/boogie-backend-v2/src/prelude/vector-array-intern-theory.bpl [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+axiom {:ctor "Vec"} (forall<T> v: Vec T :: {VecIntern(v)} VecRepIsEqual(v, VecIntern(v)));
+
+axiom {:ctor "Vec"} (forall<T> v1, v2: Vec T ::
+    {VecRepIsEqual(v1, v2)}{VecIntern(v1), VecIntern(v2)}
+    VecRepIsEqual(v1, v2) <==> VecIntern(v1) == VecIntern(v2));
+
+function VecRepIsEqual<T>(v1: Vec T, v2: Vec T): bool {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply language/move-prover/boogie-backend-v2/src/prelude/vector-array-intern-theory.bpl:27]: It's equality on the vector representation, which is not the same as the native equality `==`. We only make the API such that each function returns a projection into the Vec _ sub-domain in which native equality is the same as representation equality. But unfortunately it did not work, it times out on similar problems as SmtArrayExt does.

**Pull request merged**